src/cls/rbd/cls_rbd.cc

```diff
@@ -5767,346 +5795,1478 @@ int image_snapshot_set_copy_progress(cls_method_context_t hctx,
+    std::string group_id;
+    r = read_key(hctx, global_id_key, &group_id);
+    if (r >= 0) {
+      return -EEXIST;
+    } else if (r != -ENOENT) {
+      CLS_ERR("error reading global group id: '%s': '%s'", group_id.c_str(),
+              cpp_strerror(r).c_str());
       return r;
     }
 
-    if (!peers.empty()) {
-      CLS_ERR("mirroring peers still registered");
-      return -EBUSY;
-    }
-
-    r = remove_key(hctx, mirror::MODE);
-    if (r < 0) {
+    // make sure this was not a race for disabling
+    if (group.state == cls::rbd::MIRROR_GROUP_STATE_DISABLING) {
+      CLS_ERR("group '%s' is already disabled", group_id.c_str());
       return r;
     }
+  } else if (r < 0) {
+    CLS_ERR("error reading mirrored group '%s': '%s'", group_id.c_str(),
+	    cpp_strerror(r).c_str());
+    return r;
+  } else if (existing_group.global_group_id !=
+                group.global_group_id) {
+    // cannot change the global id
+    return -EINVAL;
+  }
 
-    r = remove_key(hctx, mirror::UUID);
-    if (r < 0) {
-      return r;
-    }
+  r = cls_cxx_map_set_val(hctx, group_key(group_id), &bl);
+  if (r < 0) {
+    CLS_ERR("error adding mirrored group '%s': %s", group_id.c_str(),
+            cpp_strerror(r).c_str());
+    return r;
+  }
+
+  bufferlist group_id_bl;
+  encode(group_id, group_id_bl);
+  r = cls_cxx_map_set_val(hctx, group_global_key(group.global_group_id),
+                          &group_id_bl);
+  if (r < 0) {
+    CLS_ERR("error adding global id for group '%s': %s", group_id.c_str(),
+            cpp_strerror(r).c_str());
+    return r;
   }
   return 0;
 }
 
-/**
- * Input:
- * @param unique peer site name (std::string)
- * @param mirror_uuid (std::string)
- * @param direction (MirrorPeerDirection) -- future use
- *
- * Output:
- * @returns 0 on success, negative error code on failure
- */
-int mirror_peer_ping(cls_method_context_t hctx, bufferlist *in,
-                     bufferlist *out) {
-  std::string site_name;
-  std::string mirror_uuid;
-  cls::rbd::MirrorPeerDirection mirror_peer_direction;
-  try {
-    auto it = in->cbegin();
-    decode(site_name, it);
-    decode(mirror_uuid, it);
+int group_status_remove(cls_method_context_t hctx,
+			const string &global_group_id);
 
-    uint8_t direction;
-    decode(direction, it);
-    mirror_peer_direction = static_cast<cls::rbd::MirrorPeerDirection>(
-      direction);
-  } catch (const ceph::buffer::error &err) {
-    return -EINVAL;
+int group_remove(cls_method_context_t hctx, const string &group_id) {
+  bufferlist bl;
+  cls::rbd::MirrorGroup group;
+  int r = group_get(hctx, group_id, &group);
+  if (r < 0) {
+    if (r != -ENOENT) {
+      CLS_ERR("error reading mirrored group '%s': '%s'", group_id.c_str(),
+	      cpp_strerror(r).c_str());
+    }
+    return r;
   }
 
-  if (mirror_peer_direction != cls::rbd::MIRROR_PEER_DIRECTION_TX) {
-    return -EINVAL;
+  /* disable_group/group_remove at the time of force promote while the remote is alive
```

Comment:
Note to self: follow up on this later.

Reply:
This causes a test in TestClsRbd to fail. The test passes on uncommenting the code. It seems logical to prevent the removal of a group unless it is has been disabled so I think this code should be enabled.